src/main/kotlin/org/pkl/lsp/VirtualFile.kt

```diff
@@ -148,51 +184,69 @@ class HttpsFile(override val uri: URI, override val project: Project) : VirtualF
   }
 }
 
-private fun ensureJarFileSystem(uri: URI) {
-  try {
-    FileSystems.newFileSystem(uri, HashMap<String, Any>())
-  } catch (e: FileSystemAlreadyExistsException) {
-    FileSystems.getFileSystem(uri)
-  }
-}
-
-class JarFile : VirtualFile {
-  val path: Path
-  override val uri: URI
-  override val project: Project
+class JarFile(override val path: Path, override val uri: URI, override val project: Project) :
+  BaseFile() {
   override val name: String
     get() = path.name
 
+  override val pklProject: PklProject? = null
+  override val pklProjectDir: VirtualFile? = null
+
+  override val `package`: PackageDependency?
+    get() =
+      project.cachedValuesManager.getCachedValue("JarFile.package(${uri})") {
+        val jarFile: Path = Path.of(URI(uri.toString().drop(4).substringBefore("!/")))
+        val jsonFile =
+          jarFile.parent.resolve(jarFile.nameWithoutExtension + ".json")
+            ?: return@getCachedValue null
+        if (!Files.exists(jsonFile)) return@getCachedValue null
+        val metadata = PackageMetadata.load(jsonFile)
+        val packageUri = metadata.packageUri
+        CachedValue(packageUri.asPackageDependency(null))
+      }
+
   override val pklAuthority: String
     get() = Origin.JAR.name.lowercase()
 
-  constructor(path: Path, uri: URI, project: Project) {
-    this.path = path
-    this.uri = uri
-    this.project = project
-  }
-
-  constructor(uri: URI, project: Project) {
-    ensureJarFileSystem(uri)
-    this.uri = uri
-    this.path = Path.of(uri)
-    this.project = project
-  }
+  override fun parent(): VirtualFile? = path.parent?.let { project.virtualFileManager.get(it) }
 
-  override fun parent(): VirtualFile? = path.parent?.let { JarFile(it, it.toUri(), project) }
-
-  override fun resolve(path: String): VirtualFile? {
-    val resolvedPath = this.path.resolve(path)
-    return if (Files.exists(resolvedPath)) JarFile(this.path.resolve(path).toUri(), project)
-    else null
-  }
+  override fun resolve(path: String): VirtualFile? =
+    project.virtualFileManager.get(this.path.resolve(path))
 
   override fun toModule(): PklModule? {
-    return Builder.findModuleInCache(uri) ?: Builder.pathToModule(path, this)
+    val builder = project.builder
+    return builder.findModuleInCache(uri) ?: builder.pathToModule(path, this)
   }
 
   override fun contents(): String =
     project.cachedValuesManager.getCachedValue("${javaClass.simpleName}-contents-${uri}") {
       CachedValue(path.readText())
     }!!
 }
+
+class EphemeralFile(private val text: String, override val project: Project) : BaseFile() {
```

Comment:
What's the purpose of this class?

Reply:
This is used by `PklNodeFactory` to create one-off modules.